File: app/src/main/java/data/room/WanicchouDatabase.java
package data.room;

import android.arch.persistence.room.Database;
import android.arch.persistence.room.Room;
import android.arch.persistence.room.RoomDatabase;
import android.arch.persistence.room.TypeConverter;
import android.arch.persistence.room.TypeConverters;
import android.content.Context;

import data.room.rel.RelatedWordDao;
import data.room.rel.RelatedWordEntity;
import data.room.voc.VocabularyDao;
import data.room.voc.VocabularyEntity;

// TODO: What's a schema export for?
@Database(
        entities = {VocabularyEntity.class,
                RelatedWordEntity.class},
        version = 1,
        exportSchema = false
)
@TypeConverters(
        {Converters.class}
)
/**
 * Database object using the Room Persistence Library.
 * Singleton design to avoid multiple instances of database
 * connection when it is not needed.
 */
public abstract class WanicchouDatabase extends RoomDatabase {
    public abstract VocabularyDao vocabularyDao();
    public abstract RelatedWordDao relatedWordDao();

    // Singleton to avoid multiple DB connections
    private static WanicchouDatabase INSTANCE;
    public static WanicchouDatabase getDatabase(final Context context){
        if (INSTANCE == null){
            synchronized (WanicchouDatabase.class){
                if(INSTANCE == null){
                    INSTANCE = Room.databaseBuilder(context.getApplicationContext(),
                            WanicchouDatabase.class, "WanicchouDatabase").build();
                }
            }
        }
        return INSTANCE;
    }
}

File: app/src/main/java/data/room/WanicchouDatabase.kt
package data.room

import android.arch.persistence.db.SupportSQLiteDatabase
import android.content.Context
import android.arch.persistence.room.*
import android.arch.persistence.room.migration.Migration
import data.arch.util.SingletonHolder
import data.room.dao.*
import data.room.entity.*

/**
 * Database object using the Room Persistence Library.
 * Singleton design to avoid multiple instances of database
 * connection when it is not needed.
 */
@Database(
        entities = [
            Vocabulary::class,
            Definition::class,
            DefinitionNote::class,
            Dictionary::class,
            Tag::class,
            VocabularyNote::class,
            VocabularyRelation::class,
            VocabularyTag::class
        ],
        version = 2
)
@TypeConverters(Converters::class)
abstract class WanicchouDatabase : RoomDatabase() {
    abstract fun definitionDao(): DefinitionDao
    abstract fun definitionNoteDao(): DefinitionNoteDao
    abstract fun dictionaryDao(): DictionaryDao
    abstract fun tagDao(): TagDao
    abstract fun vocabularyDao(): VocabularyDao
    abstract fun vocabularyNoteDao(): VocabularyNoteDao
    abstract fun vocabularyRelationDao(): VocabularyRelationDao
    abstract fun vocabularyTagDao(): VocabularyTagDao

    companion object : SingletonHolder<WanicchouDatabase, Context>({
        val MIGRATION_1_2 = object : Migration(1, 2){
            override fun migrate(database: SupportSQLiteDatabase) {
                database.execSQL(WanicchouMigration.MIGRATION_1_2_QUERY)
            }
        }
        Room.databaseBuilder<WanicchouDatabase>(it.applicationContext,
                                                        WanicchouDatabase::class.java,
                                                       "WanicchouDatabase")
                .addMigrations(MIGRATION_1_2)
                .build()
    })
}
